File: notale-v2/core/theme.py
"""Native CSS inspection and local input copying. No theme schema or compiler."""
from __future__ import annotations

import html
import os
import re
import shutil
import tempfile
from pathlib import Path
from urllib.parse import quote, unquote, urlsplit

import tinycss2 as cssparser

INTERFACE = re.compile(r"/\*\s*==== INTERFACE ====(.*?)==== /INTERFACE ====\s*\*/", re.S)
SURFACE = re.compile(r"--(?:surface|panel|card|board|tile|box|chip|well|sheet|plate)(?:-|$)", re.I)
IMAGES = {'.png', '.jpg', '.jpeg', '.webp', '.gif', '.svg'}
FONTS = {'.woff', '.woff2', '.ttf', '.otf'}


def walk(nodes):
    for node in nodes:
        yield node
        for attr in ('prelude', 'content', 'arguments', 'value'):
            children = getattr(node, attr, None)
            if isinstance(children, list):
                yield from walk(children)


def urls(nodes):
    for node in walk(nodes):
        if node.type == 'url':
            yield node, node.value
        elif node.type == 'function' and node.lower_name == 'url':
            args = [x for x in node.arguments if x.type not in ('whitespace', 'comment')]
            if len(args) != 1 or args[0].type != 'string':
                raise ValueError('url() 必须是本地文件或 fragment')
            yield node, args[0].value
        elif node.type == 'function' and node.lower_name in ('image-set', '-webkit-image-set'):
            # Quoted image-set sources need the same checks as url(), not a bypass.
            for child in node.arguments:
                if child.type == 'string':
                    yield child, child.value


def local_url(value: str, origin: Path, boundary: Path) -> tuple[Path | None, str]:
    u = urlsplit(value)
    if value.startswith('#'):
        return None, value
    if u.scheme or u.netloc or u.query or not u.path or '\\' in value or u.path.startswith('/'):
        raise ValueError(f'只支持本地相对资源及 fragment: {value}')
    path = (origin / unquote(u.path)).resolve()
    if not path.is_relative_to(boundary.resolve()):
        raise ValueError(f'资源越界: {value}')
    if not path.is_file():
        raise ValueError(f'缺少资源: {path}')
    if path.suffix.lower() not in IMAGES | FONTS:
        raise ValueError(f'不支持的资源类型: {path.name}')
    return path, ('#' + u.fragment if u.fragment else '')


def rules_of(nodes):
    for node in nodes:
        if node.type == 'qualified-rule':
            yield node
        elif node.type == 'at-rule' and node.content is not None and node.lower_at_keyword in (
                'media', 'supports', 'layer', 'container', 'scope', 'starting-style'):
            yield from rules_of(cssparser.parse_rule_list(node.content, skip_whitespace=True, skip_comments=True))


def selector_branches(prelude):
    """Split only top-level commas; functions and attribute blocks are CSS tokens."""
    branch = []
    for token in prelude:
        if token.type == 'literal' and token.value == ',':
            yield cssparser.serialize(branch).strip()
            branch = []
        else:
            branch.append(token)
    yield cssparser.serialize(branch).strip()


def inspect(css: str) -> tuple[list[str], list, str]:
    bad = []
    interface = INTERFACE.search(css)
    if not interface or not css.lstrip().startswith('/*'):
        bad.append('缺少完整的 INTERFACE 接口块')
    if not css.strip() or css.lstrip().startswith('```'):
        bad.append('主题必须是非空纯 CSS，不能有代码围栏')
    # CSS parsers recover unterminated blocks at EOF; reject that recovery explicitly.
    bare = re.sub(r'/\*.*?\*/|"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\'', '', css, flags=re.S)
    stack = []
    for char in bare:
        if char in '([{':
            stack.append(char)
        elif char in ')]}':
            if not stack or stack.pop() != dict(zip(')]}', '([{'))[char]:
                bad.append('CSS 括号不匹配')
                break
    if stack or '/*' in bare:
        bad.append('CSS 块/注释未闭合')
    nodes = cssparser.parse_stylesheet(css, skip_whitespace=True, skip_comments=True)
    for node in walk(nodes):
        if node.type == 'error':
            bad.append(f'CSS 语法错误 {node.source_line}:{node.source_column}: {node.message}')
        if node.type == 'at-rule' and node.lower_at_keyword == 'import':
            bad.append('不支持 @import；资源必须本地自包含')
    selectors, defined, defaults = [], set(), set()
    for rule in rules_of(nodes):
        selector = cssparser.serialize(rule.prelude).strip()
        selectors.append(selector)
        branches = list(selector_branches(rule.prelude))
        for branch in branches:
            if not re.search(r'\.nt-[\w-]+|#stage|:root|\bhtml\b|\bbody\b', branch):
                bad.append(f'共享样式须挂具名 .nt-*，不全局重写: {branch.strip()}')
        declarations = cssparser.parse_blocks_contents(rule.content, skip_whitespace=True, skip_comments=True)
        for d in declarations:
            if d.type == 'error':
                bad.append(f'CSS 声明错误 {d.source_line}: {d.message}')
            if d.type != 'declaration':
                continue
            defined.add(d.name)
            if any(s in (':root', 'html') for s in branches):
                defaults.add(d.name)
            # Any selector branch without a named local class publishes at shared scope.
            public = any(not re.search(r'\.nt-[\w-]+', s) for s in branches)
            if public and SURFACE.match(d.name):
                bad.append(f'不许定义承载面 token 于公共作用域: {d.name}')
            if any(re.search(r'#stage\s*$', s) for s in branches) and d.lower_name in (
                    'transform', 'transform-origin', 'position', 'left', 'top', 'width', 'height', 'overflow'):
                bad.append(f'不覆盖底盘 #stage 的 {d.name}')
    contract = interface.group(1) if interface else ''
    for name in ('--bg', '--text', '--font-sans'):
        if name not in defaults:
            bad.append(f'缺少必需 token: {name}')
    for name in variants(contract):
        if not any(re.search(r'data-variant\s*=\s*[\"\']?' + re.escape(name) + r'(?:[\"\']|\])', s) for s in selectors):
            bad.append(f'接口 variant 未定义: {name}')
    public_tokens = [name for line in contract.splitlines() if re.match(r'\s*\*?\s*token\s+', line)
                     for name in re.findall(r'--[\w-]+', line)]
    for name in public_tokens:
        if name not in defined:
            bad.append(f'接口 token 未定义: {name}')
        if SURFACE.match(name):
            bad.append(f'接口块不发布公共承载面 token: {name}')
    # Prose may describe materials or prohibit panels; keywords cannot infer permission.
    # The public SURFACE names above remain checked; broader semantics belong in guidance.
    for name in re.findall(r'^\s*\*?\s*class\s+(\.nt-[\w-]+)', contract, re.M):
        if not any(re.search(re.escape(name) + r'(?![\w-])', s) for s in selectors):
            bad.append(f'接口 class 未定义: {name}')
    return list(dict.fromkeys(bad)), nodes, contract


def variants(contract: str) -> list[str]:
    return re.findall(r'^\s*\*?\s*variant\s+([\w-]+)', contract, re.M)


REFERENCE = re.compile(r'^\s*\*?\s*reference(?:\s*:\s*|\s+)(style|user):(.+?)\s*$', re.M)


def references(css: str) -> list[tuple[str, str]]:
    interface = INTERFACE.search(css)
    return list(dict.fromkeys(REFERENCE.findall(interface.group(1)))) if interface else []


def reference_images(css: str, assets: Path) -> list[dict]:
    """Resolve declared input pictures, not CSS backgrounds or arbitrary file reads."""
    from . import style_catalog
    images = []
    for kind, value in references(css):
        if kind == 'style':
            path = style_catalog.read_path(value)
        else:
            path, _ = local_url(value, assets, assets)
            if path is None or path.suffix.lower() not in IMAGES - {'.svg'}:
                raise ValueError(f'主题参考须为本地位图: {value}')
        images.append({'id': f'{kind}:{value}', 'shot': str(path)})
    return images


def validate(css: str, assets: Path, *, browser=True) -> list[str]:
    bad, nodes, contract = inspect(css)
    try:
        resources = [local_url(value, assets, assets.parent) for _, value in urls(nodes)]
    except ValueError as exc:
        bad.append(str(exc))
        resources = []
    if bad or not browser:
        return bad
    return browser_check(css, assets, variants(contract), resources)


def browser_check(css: str, assets: Path, names: list[str], resources: list) -> list[str]:
    """Small offline probe, not an automatically authored component showcase."""
    from playwright.sync_api import sync_playwright
    bad = []
    assets.mkdir(parents=True, exist_ok=True)
    with tempfile.TemporaryDirectory(prefix='.theme-check-', dir=assets) as scratch:
        probe = Path(scratch)
        # Keep URL origin at assets, even though the isolated candidate is temporary.
        base = assets / 'base.css'
        if not base.is_file():
            base = probe / 'base.css'
            base.write_text((Path(__file__).resolve().parents[1] / 'vendor/chassis/base.css').read_text())
        (probe / 'index.html').write_text('<!doctype html><base href="' + html.escape(assets.as_uri() + '/') +
            '"><link rel="stylesheet" href="' + base.as_uri() + '"><link rel="stylesheet" href="' +
            (probe / 'candidate.css').as_uri() + '"><div id="stage"><span id="probe">中文 Ag</span>' +
            '<button id="control" class="nt-btn">操作</button><span id="hidden" class="nt-title" hidden>隐藏</span></div>', encoding='utf-8')
        # Rebase only the isolated probe; the delivered CSS stays relative and native.
        parsed = cssparser.parse_stylesheet(css)
        for token, value in urls(parsed):
            path, fragment = local_url(value, assets, assets.parent)
            if path:
                replace_url(token, path.as_uri() + fragment)
        (probe / 'candidate.css').write_text(cssparser.serialize(parsed), encoding='utf-8')
        with sync_playwright() as p:
            browser = p.chromium.launch(headless=True)
            page = browser.new_page()
            def route(request):
                u = urlsplit(request.request.url)
                path = Path(unquote(u.path)).resolve()
                if u.scheme == 'file' and path.is_relative_to(assets.parent.resolve()):
                    request.continue_()
                else:
                    request.abort()
            page.route('**/*', route)
            page.goto((probe / 'index.html').as_uri())
            # Leave compatibility branches and optional declarations to native CSS
            # fallback. Only required resolved tokens and chassis behavior gate delivery.
            no_image = False
            for name in ['', *names]:
                errors = page.evaluate('''(name) => {
                  const root=document.documentElement;
                  if(name) root.dataset.variant=name; else delete root.dataset.variant;
                  const s=getComputedStyle(root), errors=[];
                  for(const [token,prop] of [['--bg','background-color'],['--text','color'],['--font-sans','font-family']]) {
                    const value=s.getPropertyValue(token).trim();
                    if(!value || !CSS.supports(prop,value)) errors.push(`${name||'默认'}: ${token} 缺失或不可用 (${value})`);
                  }
                  return errors;
                }''', name)
                bad.extend(errors)
                await_frame = '() => new Promise(resolve=>requestAnimationFrame(()=>requestAnimationFrame(resolve)))'
                page.evaluate(await_frame)
                facts = page.evaluate('''() => {const stage=document.querySelector('#stage'),s=getComputedStyle(stage);
                  return {position:s.position,transform:s.transform,overflow:s.overflow,
                    hidden:getComputedStyle(document.querySelector('#hidden')).display,
                    brand:getComputedStyle(stage,'::after').pointerEvents,background:s.backgroundImage};}''')
                if facts['position'] != 'absolute' or facts['transform'] == 'none' or facts['overflow'] != 'hidden':
                    bad.append(f'{name or "默认"}: 覆盖了舞台定位/缩放/裁切机制')
                if facts['hidden'] != 'none' or facts['brand'] != 'none':
                    bad.append(f'{name or "默认"}: 覆盖 hidden 或品牌指针机制')
                no_image |= 'url(' not in facts['background']
            if not no_image:
                bad.append('图片主题必须提供默认或声明的无图状态')
            # Decode every referenced image, including images hidden by inactive variants.
            for path, fragment in resources:
                if path and path.suffix.lower() in IMAGES:
                    ok = page.evaluate('''async url => {let im=new Image();im.src=url;try{await im.decode();return im.naturalWidth>0}catch{return false}}''', path.as_uri() + fragment)
                    if not ok:
                        bad.append(f'图片无法加载: {path.name}')
                elif path:
                    ok = page.evaluate('''async url => {try{const f=new FontFace('probe-font',`url("${url}")`);await f.load();return true}catch{return false}}''', path.as_uri())
                    if not ok:
                        bad.append(f'字体无法加载: {path.name}')
            browser.close()
    return bad


def replace_url(token, value):
    encoded = '"' + value.replace('\\', '\\\\').replace('"', '\\"') + '"'
    if token.type == 'url':
        token.value, token.representation = value, 'url(' + encoded + ')'
    elif token.type == 'string':
        token.value, token.representation = value, encoded
    else:
        token.arguments = cssparser.parse_component_value_list(encoded)


def check_options(template, style, enabled=True):
    if style is not None and not style.strip():
        raise ValueError('--style 不得为空；修改成品主题请明确填写要求')
    if (template is not None or style is not None) and not enabled:
        raise ValueError('--template/--style 与 --no-style-director 冲突')
    if template is not None and not Path(template).exists():
        raise ValueError(f'--template 不存在: {template}')


def import_input(source: Path | None, assets: Path, *, allow_repair=False, request='') -> tuple[str, list[Path]]:
    if source is None:
        return '', []
    source = source.resolve()
    boundary = source if source.is_dir() else source.parent
    target = assets / 'style'
    def copy(path):
        resolved = path.resolve()
        if not resolved.is_relative_to(boundary):
            raise ValueError(f'资源越界/外部软链接: {path}')
        dest = target / path.relative_to(boundary)
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(resolved, dest)
        return dest
    if source.is_file():
        if source.suffix.lower() not in IMAGES - {'.svg'}:
            raise ValueError('--template 文件须为参考图片；CSS 请放在主题目录')
        from PIL import Image
        with Image.open(source) as im:
            im.verify()
        dest = target / 'shots' / source.name
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(source, dest)
        return '', [dest]
    shots = [copy(p) for p in sorted((source / 'shots').rglob('*'))
             if p.is_file() and p.suffix.lower() in IMAGES - {'.svg'}]
    # Extra package resources are copied only when explicitly named in the request;
    # no resource manifest, semantic role guessing, package scripts or blanket copy.
    for path in source.rglob('*'):
        relative = path.relative_to(source).as_posix()
        if path.is_file() and path.suffix.lower() in IMAGES | FONTS and relative in request:
            copy(path)
    theme_file = source / 'theme.css'
    if not theme_file.is_file():
        if not shots:
            raise ValueError('目录缺 theme.css 且 shots/ 内没有参考图')
        return '', shots
    if not theme_file.resolve().is_relative_to(boundary):
        raise ValueError('theme.css 外部软链接越界')
    css = theme_file.read_text(encoding='utf-8')
    errors = inspect(css)[0]
    if errors and not allow_repair:
        raise ValueError('导入主题无效（未授权重写）: ' + '；'.join(errors))
    nodes = cssparser.parse_stylesheet(css)
    for token, value in urls(nodes):
        path, fragment = local_url(value, source, boundary)
        if path:
            dest = copy(path)
            if path.suffix.lower() in FONTS:
                # Reused self-contained themes must keep redistribution notices.
                # Only adjacent font notices, not arbitrary package documents/scripts.
                for notice in path.parent.iterdir():
                    if notice.is_file() and (notice.name.upper() in ('OFL.TXT', 'LICENSE', 'LICENSE.TXT',
                            'LICENSE-OFL', 'NOTICE', 'NOTICE.TXT', 'COPYING')):
                        copy(notice)
                notices = path.parent / 'LICENSES'
                if notices.is_dir():
                    for notice in notices.rglob('*.txt'):
                        if notice.is_file():
                            copy(notice)
            replace_url(token, quote(os.path.relpath(dest, assets), safe='/') + fragment)
    imported_css = cssparser.serialize(nodes)
    # A reference is an input picture, so it is not among parsed CSS url() tokens.
    # Relocate only explicitly declared local references through the same sandbox.
    interface = INTERFACE.search(imported_css)
    if interface:
        def relocate(match):
            kind, value = match.groups()
            if kind == 'style':
                return match.group(0)
            path, _ = local_url(value, source, boundary)
            if path is None or path.suffix.lower() not in IMAGES - {'.svg'}:
                raise ValueError(f'主题参考须为本地位图: {value}')
            return 'reference user:' + copy(path).relative_to(assets).as_posix()
        block = REFERENCE.sub(relocate, interface.group(0))
        imported_css = imported_css[:interface.start()] + block + imported_css[interface.end():]
    return imported_css, shots


def publish(css: str, target: Path):
    """Only publish a complete checked candidate; never leave a partial theme on I/O failure."""
    target.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(mode='w', encoding='utf-8', dir=target.parent,
                                     prefix='.theme-publish-', suffix='.css', delete=False) as f:
        temp = Path(f.name)
        try:
            f.write(css)
            f.flush()
            os.fsync(f.fileno())
        except BaseException:
            temp.unlink(missing_ok=True)
            raise
    try:
        os.replace(temp, target)
    finally:
        temp.unlink(missing_ok=True)

File: notale-ts/src/core/theme.ts
/** Native CSS contracts and local resources from core/theme.py. */
import { existsSync, statSync } from 'node:fs';
import { mkdir, mkdtemp, writeFile, rename, rm, open } from 'node:fs/promises';
import path from 'node:path';
import { isIP } from 'node:net';
import { cssComponents, cssRules, cssBlocks, serializeCss, type CssComponent } from './css-tokenizer.js';
import { stripText, splitLines, PYTHON_SPACE } from './text.js';
import { resolvePath } from './planner-contract.js';

export const valueError = (message: string): Error => Object.assign(new Error(message), { name: 'ValueError' });

export const INTERFACE = new RegExp(`/\\*${PYTHON_SPACE}*==== INTERFACE ====(.*?)==== /INTERFACE ====${PYTHON_SPACE}*\\*/`, 'su');
const SURFACE = /^--(?:[sſ]urface|panel|card|board|t[iİı]le|box|ch[iİı]p|well|[sſ]heet|plate)(?:-|$)/iu;
// Python re.M anchors only at LF; its dot includes CR and Unicode line separators.
const LINE_START = '(?:^|(?<=\\n))';
const LINE_END = '(?=\\n|$)';
export const IMAGES = new Set(['.png', '.jpg', '.jpeg', '.webp', '.gif', '.svg']);
export const FONTS = new Set(['.woff', '.woff2', '.ttf', '.otf']);
const escapeRegex = (value: string) => value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
export const variants = (contract: string): string[] => [...contract.matchAll(new RegExp(`${LINE_START}${PYTHON_SPACE}*\\*?${PYTHON_SPACE}*variant${PYTHON_SPACE}+([\\p{L}\\p{N}_-]+)`, 'gu'))].map(match => match[1]!);
export const REFERENCE = new RegExp(`${LINE_START}${PYTHON_SPACE}*\\*?${PYTHON_SPACE}*reference(?:${PYTHON_SPACE}*:${PYTHON_SPACE}*|${PYTHON_SPACE}+)(style|user):([^\\n]+?)${PYTHON_SPACE}*${LINE_END}`, 'gu');
export function references(css: string): Array<[string, string]> {
  const contract = css.match(INTERFACE)?.[1] ?? '';
  const rows = [...contract.matchAll(REFERENCE)].map(match => [match[1]!, match[2]!] as [string, string]);
  return [...new Map(rows.map(row => [JSON.stringify(row), row])).values()];
}
function* walkCss(nodes: CssComponent[]): Generator<CssComponent> {
  for (const node of nodes) {
    yield node;
    for (const children of [node.prelude, node.content, node.arguments, node.value]) if (Array.isArray(children)) yield* walkCss(children);
  }
}
function* rulesOf(nodes: CssComponent[]): Generator<CssComponent> {
  for (const node of nodes) {
    if (node.type === 'qualified-rule') yield node;
    else if (node.type === 'at-rule' && node.content && ['media', 'supports', 'layer', 'container', 'scope', 'starting-style'].includes(node.at_keyword!.toLowerCase())) yield* rulesOf(cssRules(node.content, false));
  }
}
export class CssRoot {
  constructor(readonly nodes: CssComponent[]) {}
  toString(): string { return serializeCss(this.nodes); }
}
export function parseCss(css: string): CssRoot {
  return new CssRoot(cssRules(css, true, false, false));
}
/** Preserve empty branches and split only outside CSS component blocks. */
function selectorBranches(selector: string): string[] {
  const branches: string[] = [];
  let branch: ReturnType<typeof cssComponents> = [];
  for (const token of cssComponents(selector)) {
    if (token.type === 'literal' && token.value === ',') {
      branches.push(stripText(serializeCss(branch))); branch = [];
    } else branch.push(token);
  }
  branches.push(stripText(serializeCss(branch)));
  return branches;
}

export function inspect(css: string): { bad: string[]; root?: CssRoot; contract: string } {
  const bad: string[] = [], matched = css.match(INTERFACE);
  if (!matched || !stripText(css).startsWith('/*')) bad.push('缺少完整的 INTERFACE 接口块');
  if (!stripText(css) || stripText(css).startsWith('```')) bad.push('主题必须是非空纯 CSS，不能有代码围栏');
  const bare = css.replace(/\/\*.*?\*\/|"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*'/gs, '');
  const stack: string[] = [];
  for (const char of bare) {
    if ('([{'.includes(char)) stack.push(char);
    else if (')]}'.includes(char) && stack.pop() !== ({ ')': '(', ']': '[', '}': '{' } as Record<string, string>)[char]) { bad.push('CSS 括号不匹配'); break; }
  }
  if (stack.length || bare.includes('/*')) bad.push('CSS 块/注释未闭合');
  const contract = matched?.[1] ?? '';
  const nodes = cssRules(css);
  for (const node of walkCss(nodes)) {
    if (node.type === 'error') bad.push(`CSS 语法错误 ${node.source_line}:${node.source_column}: ${node.message}`);
    if (node.type === 'at-rule' && node.at_keyword!.toLowerCase() === 'import') bad.push('不支持 @import；资源必须本地自包含');
  }
  const selectors: string[] = [], defined = new Set<string>(), defaults = new Set<string>();
  for (const rule of rulesOf(nodes)) {
    const selector = stripText(serializeCss(rule.prelude!));
    selectors.push(selector);
    const branches = selectorBranches(selector);
    for (const branch of branches) if (!/\.nt-[\p{L}\p{N}_-]+|#stage|:root|(?<![\p{L}\p{N}_])(?:html|body)(?![\p{L}\p{N}_])/u.test(branch)) bad.push(`共享样式须挂具名 .nt-*，不全局重写: ${stripText(branch)}`);
    for (const declaration of cssBlocks(rule.content!)) {
      if (declaration.type === 'error') bad.push(`CSS 声明错误 ${declaration.source_line}: ${declaration.message}`);
      if (declaration.type !== 'declaration') continue;
      const name = declaration.name!;
      defined.add(name);
      if (branches.some(branch => [':root', 'html'].includes(branch))) defaults.add(name);
      if (branches.some(branch => !/\.nt-[\p{L}\p{N}_-]+/u.test(branch)) && SURFACE.test(name)) bad.push(`不许定义承载面 token 于公共作用域: ${name}`);
      if (branches.some(branch => new RegExp(`#stage${PYTHON_SPACE}*$`, 'u').test(branch)) && ['transform', 'transform-origin', 'position', 'left', 'top', 'width', 'height', 'overflow'].includes(name.toLowerCase())) bad.push(`不覆盖底盘 #stage 的 ${name}`);
    }
  }
  for (const name of ['--bg', '--text', '--font-sans']) if (!defaults.has(name)) bad.push(`缺少必需 token: ${name}`);
  for (const name of variants(contract)) if (!selectors.some(selector => new RegExp(`data-variant${PYTHON_SPACE}*=${PYTHON_SPACE}*["']?${escapeRegex(name)}(?:["']|\\])`, 'u').test(selector))) bad.push(`接口 variant 未定义: ${name}`);
  for (const line of splitLines(contract)) if (new RegExp(`^${PYTHON_SPACE}*\\*?${PYTHON_SPACE}*token${PYTHON_SPACE}+`, 'u').test(line)) for (const name of line.match(/--[\p{L}\p{N}_-]+/gu) ?? []) {
    if (!defined.has(name)) bad.push(`接口 token 未定义: ${name}`);
    if (SURFACE.test(name)) bad.push(`接口块不发布公共承载面 token: ${name}`);
  }
  for (const match of contract.matchAll(new RegExp(`${LINE_START}${PYTHON_SPACE}*\\*?${PYTHON_SPACE}*class${PYTHON_SPACE}+(\\.nt-[\\p{L}\\p{N}_-]+)`, 'gu'))) if (!selectors.some(selector => new RegExp(escapeRegex(match[1]!) + '(?![\\p{L}\\p{N}_-])', 'u').test(selector))) bad.push(`接口 class 未定义: ${match[1]}`);
  return { bad: [...new Set(bad)], root: new CssRoot(nodes), contract };
}
export function decodeCss(text: string): string {
  return text.replace(/\\([0-9a-f]{1,6})(?:\r\n|[\n\r\t\f ])?|\\([^\n\r\f])/gi, (_all, hex, escaped) => {
    if (!hex) return escaped;
    const code = parseInt(hex, 16);
    return String.fromCodePoint(code === 0 || code > 0x10ffff || (code >= 0xd800 && code <= 0xdfff) ? 0xfffd : code);
  });
}
export interface CssUrl { value: string; replace(value: string): void }
export function cssUrls(root: CssRoot): CssUrl[] {
  const result: CssUrl[] = [];
  const append = (node: CssComponent, value: string) => result.push({ value, replace(value) {
    const encoded = '"' + value.replaceAll('\\', '\\\\').replaceAll('"', '\\"') + '"';
    if (node.type === 'url') { node.value = value; node.representation = 'url(' + encoded + ')'; }
    else if (node.type === 'string') { node.value = value; node.representation = encoded; }
    else node.arguments = cssComponents(encoded);
  } });
  for (const node of walkCss(root.nodes)) {
    if (node.type === 'url') append(node, String(node.value));
    else if (node.type === 'function' && node.name!.toLowerCase() === 'url') {
      const args = node.arguments!.filter(token => token.type !== 'whitespace' && token.type !== 'comment');
      if (args.length !== 1 || args[0]!.type !== 'string') throw Object.assign(new Error('url() 必须是本地文件或 fragment'), { name: 'ValueError' });
      append(node, String(args[0]!.value));
    } else if (node.type === 'function' && ['image-set', '-webkit-image-set'].includes(node.name!.toLowerCase())) {
      for (const child of node.arguments!) if (child.type === 'string') append(child, String(child.value));
    }
  }
  return result;
}
export function isWithin(file: string, boundary: string): boolean {
  const relative = path.relative(resolvePath(boundary), resolvePath(file));
  return relative !== '..' && !relative.startsWith(`..${path.sep}`) && !path.isAbsolute(relative);
}
export const unquoteUrl = (value: string): string => value.replace(/(?:%[0-9a-f]{2})+/gi, bytes => Buffer.from(bytes.replaceAll('%', ''), 'hex').toString('utf8'));
/** String URL splitting and netloc validation used by Python resource tools. */
export function splitResourceUrl(value: string): { scheme: string; netloc: string; path: string; query: string; fragment: string } {
  const fail = (message: string): never => { throw Object.assign(new Error(message), { name: 'ValueError' }); };
  let rest = value.replace(/^[\x00-\x20]+/, '').replace(/[\t\r\n]/g, '');
  const prefix = rest.match(/^[a-z][a-z0-9+.-]*:/i)?.[0] ?? '';
  const scheme = prefix.slice(0, -1).toLowerCase(); rest = rest.slice(prefix.length);
  let netloc = '', query = '', fragment = '';
  if (rest.startsWith('//')) {
    const end = rest.slice(2).search(/[/?#]/), length = end < 0 ? rest.length : end + 2;
    netloc = rest.slice(2, length); rest = rest.slice(length);
    if (netloc.includes('[') !== netloc.includes(']')) fail('Invalid IPv6 URL');
    if (netloc.includes('[') && netloc.includes(']')) {
      const authority = netloc.slice(netloc.lastIndexOf('@') + 1), open = authority.indexOf('[');
      let hostname: string;
      if (open >= 0) {
        if (open > 0) fail('Invalid IPv6 URL');
        const close = authority.indexOf(']', open + 1);
        hostname = close < 0 ? authority.slice(open + 1) : authority.slice(open + 1, close);
        const port = close < 0 ? '' : authority.slice(close + 1);
        if (port && !port.startsWith(':')) fail('Invalid IPv6 URL');
      } else hostname = authority.split(':')[0]!;
      if (hostname.startsWith('v')) {
        if (!/^v[0-9a-fA-F]+\..+$/.test(hostname)) fail('IPvFuture address is invalid');
      } else {
        const [address, ...zone] = hostname.split('%');
        const family = isIP(address!);
        if (!family || (family === 4 && zone.length) || zone.length > 1 || (zone.length === 1 && !zone[0])) {
          const quote = hostname.includes("'") && !hostname.includes('"') ? '"' : "'";
          const repr = [...hostname].map(c => c === quote || c === '\\' ? '\\' + c : /[\p{C}\p{Z}]/u.test(c) && c !== ' ' ? (c.codePointAt(0)! <= 255 ? '\\x' : c.codePointAt(0)! <= 65535 ? '\\u' : '\\U') + c.codePointAt(0)!.toString(16).padStart(c.codePointAt(0)! <= 255 ? 2 : c.codePointAt(0)! <= 65535 ? 4 : 8, '0') : c).join('');
          fail(quote + repr + quote + ' does not appear to be an IPv4 or IPv6 address');
        }
        if (family === 4) fail('An IPv4 address cannot be in brackets');
      }
    }
  }
  const hash = rest.indexOf('#'); if (hash >= 0) { fragment = rest.slice(hash + 1); rest = rest.slice(0, hash); }
  const question = rest.indexOf('?'); if (question >= 0) { query = rest.slice(question + 1); rest = rest.slice(0, question); }
  const check = netloc.replace(/[@:#?]/g, ''), normalized = check.normalize('NFKC');
  if (normalized !== check && /[/?#@:]/.test(normalized)) fail("netloc '" + netloc + "' contains invalid characters under NFKC normalization");
  return { scheme, netloc, path: rest, query, fragment };
}
export function localUrl(value: string, origin: string, boundary: string): [string | null, string] {
  if (value.startsWith('#')) return [null, value];
  const { scheme, netloc, path: pathname, query, fragment } = splitResourceUrl(value);
  if (scheme || netloc || query || !pathname || value.includes('\\') || pathname.startsWith('/')) throw valueError(`只支持本地相对资源及 fragment: ${value}`);
  // urllib.unquote decodes valid percent bytes even alongside malformed escapes,
  // replacing invalid UTF-8 instead of falling back to the original entire path.
  const decoded = unquoteUrl(pathname);
  const file = resolvePath(decoded, origin);
  if (!isWithin(file, boundary)) throw valueError(`资源越界: ${value}`);
  if (!existsSync(file) || !statSync(file).isFile()) throw valueError(`缺少资源: ${file}`);
  if (!IMAGES.has(path.extname(file).toLowerCase()) && !FONTS.has(path.extname(file).toLowerCase())) throw valueError(`不支持的资源类型: ${path.basename(file)}`);
  return [file, fragment ? '#' + fragment : ''];
}
export function checkOptions(template: string | undefined, style: string | undefined): void {
  if (style !== undefined && !stripText(style)) throw valueError('--style 不得为空；修改成品主题请明确填写要求');
  if (template !== undefined && !existsSync(template)) throw valueError(`--template 不存在: ${template}`);
}
export async function publish(css: string, target: string): Promise<void> {
  await mkdir(path.dirname(target), { recursive: true });
  const scratch = await mkdtemp(path.join(path.dirname(target), '.theme-publish-'));
  const temporary = path.join(scratch, 'theme.css');
  try {
    await writeFile(temporary, css, 'utf8');
    const handle = await open(temporary, 'r+');
    try { await handle.sync(); } finally { await handle.close(); }
    await rename(temporary, target);
  } finally { await rm(scratch, { recursive: true, force: true }); }
}
